### PR TITLE
feat(portal): Add JumpCloud directory sync integration

### PR DESCRIPTION
**Pull request by bmanifold**
This PR adds JumpCloud directory sync integration.  

Along with adding JumpCloud support, it also slightly refactored the `identity_providers/new` page as shown below:

Screenshot 2024-05-08 at 11 21 16 AM

### CODE REVIEW

File: elixir/apps/web/test/web/live/settings/identity_providers/jumpcloud/connect_test.exs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+          }
+        )
+
+      assert redirected_to(conn) ==
+               ~p"/#{account}/settings/identity_providers/jumpcloud/#{provider}"
+
+      assert flash(conn, :error) == "Your session has expired, please try again."
+    end
+
+    ## TODO: Re-work this test since the API key is hard coded/long lived for JumpCloud API
```
Comment (bmanifold) [inline elixir/apps/web/test/web/live/settings/identity_providers/jumpcloud/connect_test.exs]: I just realized I forgot to update this test.  I'll update this, but I think the rest of the PR should be ready for review.

**Comment by AndrewDryga** [comment]
@bmanifold for the identity provider selection page I think we should communicate that OpenID Connect also supports Google, Okta, etc.

**Comment by bmanifold** [comment]
> @bmanifold for the identity provider selection page I think we should communicate that OpenID Connect also supports Google, Okta, etc.

Yeah, agreed.  I was trying to think through that this morning with how to integrate the other providers logos in the message too, but it was throwing off the visual alignment.  I'll play around a little more with it.

Also, I had talked with @jamilbk about this, but as I was refactoring this, I realized that what we really should do is not distinguish between AuthN with or without directory sync on this page, but rather, just have a list of providers we support (with a 'Customer OIDC' provider at the bottom) and then once they select their provider ask the user if they want to also setup directory sync, assuming they are on the proper plan of course.  However, that's a little larger refactor, so I just stuck with this for now.

File: elixir/apps/web/test/web/live/settings/identity_providers/jumpcloud/connect_test.exs (hunk 1 above)
Comment (bmanifold) [reply elixir/apps/web/test/web/live/settings/identity_providers/jumpcloud/connect_test.exs]: Updated.

**Comment by bmanifold** [comment]
@AndrewDryga @jamilbk, what do you guys think about this for adding in a note about the OIDC provider.  (Also simplified the text for each column description)

Screenshot 2024-05-08 at 11 30 16 PM

**Comment by jamilbk** [comment]
> @AndrewDryga @jamilbk, what do you guys think about this for adding in a note about the OIDC provider. (Also simplified the text for each column description)
> 
> Screenshot 2024-05-08 at 11 30 16 PM

329117104-c4a462f3-66d9-4d90-ab3b-0cae6a4ecd67

Mostly some clarification suggestions

- some wording cleanup
- link to docs for both
- show upgrade badge for non-enterprise plans, for enterprise plans maybe it makes sense to hide that sentence
- Can move the small logos inline with the text on the left? Another option could be to add them to the `OpenID Connect` box in the lower-left preceded by "Supports", like "Supports (G), (O), (E)"

**Comment by bmanifold** [comment]
@jamilbk, is this what you had in mind? (Also, I forgot to include it on the last screenshot, but the "upgrade badge" I had on each of the providers themselves.  Would you prefer to remove those and just have 1 badge where you mentioned?)

Screenshot 2024-05-09 at 10 46 37 AM

**Comment by jamilbk** [comment]
@bmanifold Got a notification for an update, but can't find the comment now, maybe it was deleted?

**Comment by bmanifold** [comment]
@jamilbk, yeah I had put up a screenshot of the changes, but then realized that I forgot some of the wording you put in.  Then after adding that wording things were really thrown off.  It may be better to talk these changes over synchronously if possible.

**Comment by bmanifold** [comment]
@jamilbk, what do you think of these:

Screenshot 2024-05-10 at 12 48 09 AM


### Clicking on a 'sync capable' provider on a non-enterprise plan
Screenshot 2024-05-10 at 12 48 39 AM

**Comment by jamilbk** [comment]
> @jamilbk, what do you think of these:

Much better! Last set of suggestions:

- Sync icon looks weird than I thought. Maybe cleaner to remove? I think it could still make sense in the actual form, prepended to a new Step titled "(icon) Step 2: Configure Directory Sync (upgrade to unlock)"
- Some wording / docs link suggestions
- Could shrink the hero buttons just a tad


329461178-9421daba-6b6a-4dd2-a966-9f743de16e15

329461176-e22d8bb6-2144-4f28-a2ff-2013118edba5

**Comment by bmanifold** [comment]
> Much better! Last set of suggestions:
> 
> 329461176-e22d8bb6-2144-4f28-a2ff-2013118edba5


Shrinking the width and adding another docs link sounds good.  As far as removing the icon from the help section, my only hesitation there is that if we don't have that icon defined somewhere on the page I'm worried people aren't going to know what that means.  What are your thoughts?

**Comment by jamilbk** [comment]
> What are your thoughts?

Edited my comment to include the second screenshot. We could simply wait to tell them about directory sync until they get to the actual form. At that point they're already halfway through setting up the connector and the context should make it obvious. In that step, we could prepend the icon to the step title: "(icon) Step 2: Configure Directory Sync"

**Comment by jamilbk** [comment]
> > What are your thoughts?
> 
> Edited my comment to include the second screenshot. We could simply wait to tell them about directory sync until they get to the actual form. At that point they're already halfway through setting up the connector and the context should make it obvious. In that step, we could prepend the icon to the step title: "(icon) Step 2: Configure Directory Sync"

Though I guess with this approach, they might complete the form, and then have to recreate the adapter, but I don't think that's a huge deal... this way they'll see exactly where they'd go to add the scopes for the directory sync connector they'd create when they have Enterprise

**Comment by bmanifold** [comment]
> > > What are your thoughts?
> > 
> > 
> > Edited my comment to include the second screenshot. We could simply wait to tell them about directory sync until they get to the actual form. At that point they're already halfway through setting up the connector and the context should make it obvious. In that step, we could prepend the icon to the step title: "(icon) Step 2: Configure Directory Sync"
> 
> Though I guess with this approach, they might complete the form, and then have to recreate the adapter, but I don't think that's a huge deal... this way they'll see exactly where they'd go to add the scopes for the directory sync connector they'd create when they have Enterprise

I'm a little confused on the grayed out part and what exactly would be in the grayed out step.  Are you wanting the exact message I have in the help section moved to Step 2 in the OIDC form and grayed out?  Or are you saying you want individual directory sync instructions for the given provider that was clicked on in Step 2 and grayed out?

Maybe we could sync up real quick in the morning to clarify?

**Comment by jamilbk** [comment]
> Are you wanting the exact message I have in the help section moved to Step 2 in the OIDC form and grayed out? 

Was just thinking some dummy text to make it a bit easier, yeah can sync in the AM

**Review by jamilbk — COMMENTED** [review]
Maybe mark this as draft while we explore SCIM?

**Comment by bmanifold** [comment]
@jamilbk, yeah, apologies, I've been meaning to either mark this as a draft or just close it.  Do you have a preference?

**Comment by jamilbk** [comment]
Superseded by #5269

**Pull request closed without merging**